### ISSUE TITLE
Add support for Node 16

### ISSUE THREAD

**Issue by mraible** [jhipster/generator-jhipster#16744]
##### **Overview of the feature request**

Node.js 16 gets promoted to long-term support (LTS) next week!

https://twitter.com/reverentgeek/status/1450501408759947267

##### **Motivation for or Use Case**

We support the latest LTS for Node.js, so there might be some work to do here. I'll add a bug bounty, but we can increase it if necessary.

##### **Related issues or PR**

- [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by RobertBurrellDonkin** [comment]
Running ```nvm test``` on  AmazonLinux2 without ```nvm use 14``` fails with ```Error [ERR_MULTIPLE_CALLBACK]: Callback called multiple times```

See #16918

Might also have been seen in the wild  https://stackoverflow.com/questions/69835351/jhipster-version-7-3-1-latest-having-error-in-node-during-the-code-generation

**Comment by mshima** [comment]
I took a look why tests are failing at node 16.
It's related to a transform stream behavior change.
It was quite hard to find and I could not create a way to reproduce the behavior.

**Comment by RobertBurrellDonkin** [comment]
> I took a look why tests are failing at node 16.
> It's related to a transform stream behavior change.
> It was quite hard to find and I could not create a way to reproduce the behavior.

Thanks for jumping in @mshima :-)

There's a possibility that I might be able to come up with a reproduction recipe but I'll need to play around on AWS to be sure...

**Comment by RobertBurrellDonkin** [comment]
On MacOSX

```jhipster jdl sample.jdl```

with

```
application {
  config {
    baseName bootifulmusic,
    applicationType monolith,
    packageName com.jhipster.demo.bootifulmusic,
    databaseType neo4j,
    devDatabaseType neo4j,
    prodDatabaseType neo4j,
    enableHibernateCache false,
    testFrameworks [protractor],
    searchEngine elasticsearch
  }
  entities *
}

entity Artist {
	name String required
}

entity Genre {
	name String required
}

entity Track {
	name String required
}

entity Album {
	name String required
}

relationship OneToOne {
	Album{artist(name)} to Artist
    Album{genre(name)} to Genre
}

relationship OneToMany {
	Album{track(name)} to Track{album(name)}
}

paginate Artist, Genre, Track, Album with pagination
search Artist with elasticsearch
```

generates after ```nvm use 14``` but fails on ```node 16``` with

```
Error running generator app: Error [ERR_MULTIPLE_CALLBACK]: Callback called multiple times
Error [ERR_MULTIPLE_CALLBACK]: Callback called multiple times
    at NodeError (node:internal/errors:371:5)
    at onFinish (node:internal/streams/writable:667:37)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
    at runNextTicks (node:internal/process/task_queues:65:3)
    at listOnTimeout (node:internal/timers:526:9)
    at processTimers (node:internal/timers:500:7) {
  code: 'ERR_MULTIPLE_CALLBACK'
}
ERROR! Callback called multiple times
Error [ERR_MULTIPLE_CALLBACK]: Callback called multiple times
    at NodeError (node:internal/errors:371:5)
    at onFinish (node:internal/streams/writable:667:37)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
    at runNextTicks (node:internal/process/task_queues:65:3)
    at listOnTimeout (node:internal/timers:526:9)
    at processTimers (node:internal/timers:500:7) {
  code: 'ERR_MULTIPLE_CALLBACK'
}
```

**Comment by mraible** [comment]
Increasing bug bounty because upgrade is not trivial.

**Comment by mshima** [comment]
The internal changes were not enough to fix Node 16 support.
Node 16 fails on unhandled rejections.
React and Vue frontends have unhandled exceptions.
React problem looks trivial.

**Comment by ksilz** [comment]
Another reason for going with Node 16 is that it’s the first node release with native Mac ARM64/M1 support. Here’s [the last Node 14 release](https://nodejs.org/download/release/v14.18.1/): It just has `*-darwin-x64*` files. Here's [the 16 one:](https://nodejs.org/download/release/v16.13.0/): It also has `*-darwin-arm64*` files.

**Comment by mraible** [comment]
@mshima Can this be closed now?

**Comment by pascalgrimaud** [comment]
@mraible : this PR https://github.com/jhipster/generator-jhipster/pull/16883 is still opened, so we should not close this

**Comment by mshima** [comment]
> @mshima Can this be closed now?

@mraible unfortunately there are 1 bug with react and 1 bug with vue that needs to be fixed before we can switch to node 16.

Just changed https://github.com/jhipster/generator-jhipster/issues/16982 title to fix unhandled exception in React.

This additional $200 should go to fix unhandled exceptions in VUE (jest test and cypress).
https://github.com/jhipster/generator-jhipster/blob/cc5fcdd73fed569276dd17cba0db0bb40a57a991/generators/cypress/templates/src/test/javascript/cypress/support/index.ts.ejs#L49-L57

**Comment by mraible** [comment]
Please accept the full bounty for this one. Then create a new issue and
let's assign 200 to that one.

**Comment by mshima** [comment]
Following https://github.com/jhipster/generator-jhipster/pull/16883#issuecomment-956416314.
We need to update or add node 16 to matrix:
- react-jhipster
- prettier-java
- jhipster-devbox
- generator-jhipster-blueprint and generator-jhipster-module (are they working?)

Blueprints should switch when switching to jhipster v7.5.0

**Comment by DanielFran** [comment]
@pascalgrimaud I believe that we should share the bug bounty between me and @mshima, and open specific issues for the remaining projects.
What do you think?

**Comment by pascalgrimaud** [comment]
Same opinion than for Spring Boot 2.6
Claim the bounty, and @mshima can do the same. No need to share as it's a lot of work here too

**Comment by DanielFran** [comment]
Thanks @pascalgrimaud Bug bounty claimed: https://opencollective.com/generator-jhipster/expenses/60929

**Comment by pascalgrimaud** [comment]
@DanielFran : approved

**Comment by emilpaw** [comment]
@pascalgrimaud This can be closed, right?

**Comment by mshima** [comment]
@pascalgrimaud bounty claimed https://opencollective.com/generator-jhipster/expenses/68439.

**Comment by pascalgrimaud** [comment]
@mshima : can you edit the title at open collective with correct ref ? It's 16744, not 16774 :-)

**Comment by mshima** [comment]
Oops.. Fixed. Thanks.

**Comment by pascalgrimaud** [comment]
@mshima : approved :-)
